tritium-lib/src/main/java/com/palantir/tritium/proxy/ByteBuddyInstrumentation.java

```diff
@@ -0,0 +1,234 @@
+            List<Method> allMethods = new ArrayList<>();
+            for (Class<?> iface : interfaces) {
+                boolean allowDirectAccess = iface.isAssignableFrom(interfaceClass);
+                // If the method is not available on our target interface, add an internal getter to cast the delegate
+                // object into the target interface.
+                if (!allowDirectAccess) {
+                    builder = builder.defineMethod(delegateMethod(iface), iface, Modifier.PRIVATE | Modifier.FINAL)
+                            .intercept(FieldAccessor.ofField("delegate")
+                                    .withAssigner(Assigner.DEFAULT, Assigner.Typing.DYNAMIC));
+                }
+                for (Method method : iface.getMethods()) {
+                    int index = allMethods.size();
+                    allMethods.add(method);
+                    // Retain tritium proxy detail where hashcode, equals, and toString cannot be instrumented.
+                    builder = builder.method(ElementMatchers.not(ElementMatchers.isHashCode()
+                            .or(ElementMatchers.isEquals())
+                            .or(ElementMatchers.isToString()))
+                            .and(ElementMatchers.is(method)))
+                            .intercept(Advice.withCustomMapping()
+                                    .bind(ByteBuddyInstrumentationAdvice.MethodIndex.class, index)
+                                    .to(ByteBuddyInstrumentationAdvice.class)
+                                    .wrap(allowDirectAccess
+                                            ? MethodDelegation.toField("delegate")
+                                            : MethodDelegation.toMethodReturnOf(delegateMethod(iface))));
+                }
+            }
+            return builder
+                    .defineField("delegate", interfaceClass,
+                            Opcodes.ACC_PRIVATE | Opcodes.ACC_FINAL)
+                    .defineField("invocationEventHandler", InvocationEventHandler.class,
+                            Opcodes.ACC_PRIVATE | Opcodes.ACC_FINAL)
+                    .defineField("instrumentationFilter", InstrumentationFilter.class,
+                            Opcodes.ACC_PRIVATE | Opcodes.ACC_FINAL)
+                    .defineField("methods", Method[].class, Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC)
+                    .initializer(new StaticFieldLoadedTypeInitializer("methods", allMethods.toArray(new Method[0])))
+                    .make()
+                    .load(classLoader, ClassLoadingStrategy.Default.WRAPPER)
+                    .getLoaded();
+        });
+    }
+
+    private static ClassLoader getClassLoader(Class<?> clazz) {
+        ClassLoader loader = clazz.getClassLoader();
+        // Some internal classes return a null classloader, in these cases we provide the instrumentation loader.
+        if (loader == null) {
+            return ByteBuddyInstrumentation.class.getClassLoader();
```

Comment:
Should we prefer thread context class-loader before this class’?

Reply:
I think we should check the context classloader, but we should prefer this one to avoid caching the same class against several child ccls.

I should add validation that the fallback classloader can actually load the provided class. It may be worthwhile checking `ByteBuddyInstrumentation.class.getClassLoader()` before the serviceInterface classloader to allow for potentially more additional interfaces to be instrumented.